### PR TITLE
Add tests for counter component and slice

### LINKED ISSUE #23: Include tests for Counter as example
Inspired by reduxjs/cra-template-redux-typescript#21

I'm starting out with integration tests using React Testing Library and Jest (without mocking the store), but I'm open to suggestions. It seems like the existing Redux examples are doing shallow rendering (not using the store) with Enzyme and direct Redux calls for unit tests, but I don't think we should suggest Enzyme or unit testing components. I might add isolated unit tests for the store code if you'd think it's useful though.

### PR DESCRIPTION
**Pull request by aldreth**
closes #23 - I also have a typescript equivalent to this pr that I can create if you're happy with this.

### CODE REVIEW

File: template.json
Comment (aldreth) [inline template.json:11]: This will be required till there's a release of CRA with jest 25 - it has been merged to master, but not released.

See `MutationObserver shim removed.` here - https://github.com/testing-library/dom-testing-library/releases/tag/v7.0.0

File: template/src/features/counter/Counter.test.js
Comment (aldreth) [inline template/src/features/counter/Counter.test.js:2]: This will be required till there's a release of CRA with jest 25 - it has been merged to master, but not released.

See `MutationObserver shim removed` here - https://github.com/testing-library/dom-testing-library/releases/tag/v7.0.0

File: template/src/features/counter/Counter.js
Comment (BenLorantfy) [inline template/src/features/counter/Counter.js]: I think we should avoid adding `data-testid`. testing-library discourages using this when it's possible to query other ways, and it would be good to encourage best practices

> In the spirit of the guiding principles, it is recommended to use this only after the other queries don't work for your use case. Using data-testid attributes do not resemble how your software is used and should be avoided if possible.
https://testing-library.com/docs/dom-testing-library/api-queries#bytestid

I think you could use `getByLabelText("Increment value")` instead

File: template/src/features/counter/Counter.js
Comment (BenLorantfy) [inline template/src/features/counter/Counter.js]: In a similar vein, but slightly more complicated, you do this to avoid `data-testid`:
```suggestion
        <span className={styles.value}>
          <span className="sr-only">Count is </span>
          {count}
        </span>
```

and then query by:`screen.getByText(/Count is \d+/i)`

You'd have to add this though: https://gist.github.com/ffoodd/000b59f431e3e64e4ce1a24d5bb36034#file-style-css-L30

File: template/src/features/counter/counterSlice.js
Comment (BenLorantfy) [inline template/src/features/counter/counterSlice.js]: I think this code is a little less complicated but get's the same point across:
```suggestion
    await delay(1000);
    thunkApi.dispatch(incrementByAmount(amount));
```

Comment (markerikson) [reply template/src/features/counter/counterSlice.js]: My main concern here is that the point of `createAsyncThunk` is to auto-dispatch actions based on a promise's lifecycle, and then handle those actions in a reducer, with the return value of this callback becoming the promise.

The hand-written example thunk wasn't making any AJAX calls, it was just showing you can have any kind of async logic at all, and have access to `dispatch`.

This use of `createAsyncThunk` doesn't fit either approach. It's not returning a payload, it's dispatching a separate action type, and it's not the right semantic use for `createAsyncThunk`.

I've got mixed opinions on including examples of `createAsyncThunk` right in the template. Biggest issue is that we don't have any API to call here.

Tagging @phryneas and @msutkowski to get their thoughts.

Comment (msutkowski) [reply template/src/features/counter/counterSlice.js]: I fully agree with these points and don't have any counters to them.

If you choose to keep the `createAsyncThunk` usage, I'd return the value and handle the case in `extraReducers` so that it more closely matches normal usage. I'd still question that implementation though.

Comment (BenLorantfy) [reply template/src/features/counter/counterSlice.js]: So I hadn't heard of `createAsyncThunk` when I made this comment. Reading up on it a bit I feel like this isn't the best showcase of `createAsyncThunk`. If the typical use-case is to make an api call, it should probably make an api call.  Otherwise people will be confused why we're using `createAsyncThunk` here.

File: template/src/features/counter/Counter.js
Comment (jihchi) [reply template/src/features/counter/Counter.js]: Or `getByRole('button', { name: 'Increment value' })`, this is a11y friendly

File: template/src/features/counter/counterSlice.js
Comment (phryneas) [reply template/src/features/counter/counterSlice.js]: Yeah. Adding example functionality that actually queries something (mocked by `msw` or similar)  might be an idea - but here, `createAsyncThunk` is not used correctly.

Comment (aldreth) [reply template/src/features/counter/counterSlice.js]: @markerikson I agree with what you're saying here, `createAsyncThunk` is not used as it would be normally. However I really like `createAsyncThunk` and wonder if it would be useful to have it used in the template.

I think there are a few options & I'm happy to do whichever you're most happy with.

* Revert the change here, don't use `createAsyncThunk`
  * I couldn't work out how to test the function that was there before at all - any tips?
  * or don't test that function
* Use  `createAsyncThunk` and add a reducer in `extraReducers` for handling `incrementAsync/resolved`
  * Use the `fakeAsyncCall` that I've added
  * Or do more elaborate api mocking

Thoughts?

File: template/src/index.css
Comment (BenLorantfy) [inline template/src/index.css]:
```suggestion
	clip: rect(1px, 1px, 1px, 1px) !important;
```

File: template/src/index.css
Comment (BenLorantfy) [inline template/src/index.css]:
```suggestion
		clip-path: inset(50%) !important;
```

File: template/src/index.css
Comment (BenLorantfy) [inline template/src/index.css]:
```suggestion
	white-space: nowrap !important;
```

File: template/src/features/counter/Counter.js
Comment (BenLorantfy) [inline template/src/features/counter/Counter.js]: The below would be ideal because right now screen readers will read the count twice.
```suggestion
          <span className="sr-only">Count is</span>
          {count}
```

But this would also work if you can't easily test the above
```suggestion
          <span className="sr-only">Count is {count}</span>
          <span aria-hidden>{count}</span>
```

File: yarn.lock
Comment (BenLorantfy) [inline yarn.lock]: IMO should remove this file since we're using `npm` to install dependencies

Comment (BenLorantfy) [reply yarn.lock]: See https://github.com/reduxjs/cra-template-redux/blob/master/package-lock.json

Comment (aldreth) [reply yarn.lock]: Oops!

File: template/src/features/counter/Counter.js
Comment (aldreth) [reply template/src/features/counter/Counter.js]: The first method you suggested was tricky to get the value, as it was in a separate element.

Comment (aldreth) [reply template/src/features/counter/Counter.js]: I've force pushed with changes that use `getByTitle`  rather than adding css - how does that look to you @BenLorantfy ?

Comment (BenLorantfy) [reply template/src/features/counter/Counter.js]: I did some research on this and I don't think `title` is recommended
> Short answer: Don’t use them, except in special circumstances.
https://a11yproject.com/posts/title-attributes/

What do you think of this: 
```
<output aria-label="Counter value" aria-live="polite">
  5
</output>
```

and then:
`screen.getByRole('status', { name: /counter value/i })`

See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/output
See: [Playground link](
https://testing-playground.com/?markup=DwewrgLgDpAECGAnAlvAtAG3gIwKYYF4AiAYXADsJdFYA3eDMXIgPgCgBWN4AenGkgsgA&query=M4YwTgphB2B0DmEAuAhAngJQPYBsIAoByYJAQyQFdhCAaAAgG87pSBbCALjoHoQsLoSCGDoA3UjgoRuASzoBfAJRA)

Comment (aldreth) [reply template/src/features/counter/Counter.js]: LGTM. Done

File: template/src/features/counter/counterSlice.js
Comment (aldreth) [reply template/src/features/counter/counterSlice.js]: @markerikson I've pushed an update that uses `createAsyncThunk` more appropriately. What do you think?

File: template/src/features/counter/counterSlice.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.js]:
```suggestion
// It allows us to perform async logic and automatically dispatches actions matching Promise life cycle states: pending, fulfilled or rejected. It can be dispatched like a regular action:
```

File: template/src/features/counter/counterSlice.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.js]:
```suggestion
```

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.test.js]:
```suggestion
    it('adds the payload to the value when incrementAsync is fulfilled', () => {
```

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.test.js:44]: A couple of thoughts here:
- This test basically just tests that `createAsyncThunk` dispatches actions in order as designed. I'm not sure if there is much value in this type of test, but I could be wrong _(like in the instance the intended purpose of this specific test is to pull back the curtain on how `createAsyncThunk` works? That being said, RTK already does this in-depth.)_
- If we're going to include these tests, it might be a good idea to check `type: incrementAsync.(fulfilled|pending|rejected).type` instead of a hardcoded string so they're less brittle - the same way as done in the reducers test above.

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.test.js]:
```suggestion
    it('subtracts one from the current value', () => {
```

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.test.js]:
```suggestion
    it('adds one to the current value', () => {
```

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.test.js]:
```suggestion
    it('adds the payload to the current value', () => {
```

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.test.js]:
```suggestion
    it('does not modify the state when incrementAsync is pending', () => {
```

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [inline template/src/features/counter/counterSlice.test.js]:
```suggestion
    it('returns the current value of the counter', () => {
```

File: template/src/features/counter/Counter.test.js
Comment (msutkowski) [inline template/src/features/counter/Counter.test.js]:
```suggestion
test('can render the Counter and have the store be updated by actions', async () => {
```

**Review by msutkowski — COMMENTED** [review]
These are just some general comments and are mostly about readability. Please feel free to disregard if they're totally off-base. @markerikson or someone else might have much better feedback :)

File: template.json
Comment (nickserv) [inline template.json:7]: I would recommend keeping `user-event` and using it as often as possible over `fireEvent` from `react-testing-library`. `user-event`'s goal is to eventually replace `fireEvent`, as `user-event` models actual user behavior in a browser more accurately rather than forcing you to manually fire the same events that a browser would.

File: template/src/features/counter/fakeApi.js
Comment (nickserv) [inline template/src/features/counter/fakeApi.js]: I know this is probably just here for demonstrational purposes, but something a little more real-world would be nice to showcode best practices. For example this could use `fetch` to call out to https://example.com/ or https://jsonplaceholder.typicode.com/.

File: template/src/features/counter/counterSlice.js
Comment (nickserv) [inline template/src/features/counter/counterSlice.js]: The result of `fakeAsyncCall` could just be returned directly if it's already a Promise.

```suggestion
 (amount, thunkApi) => fakeAsyncCall(amount)
```

File: template/src/features/counter/Counter.test.js
Comment (nickserv) [inline template/src/features/counter/Counter.test.js]: Shouldn't this be `toHaveTextContent`?

File: template/src/features/counter/Counter.test.js
Comment (nickserv) [inline template/src/features/counter/Counter.test.js]:
```suggestion
  userEvent.click(increment);
```

File: template/src/features/counter/Counter.test.js
Comment (nickserv) [inline template/src/features/counter/Counter.test.js]:
```suggestion
  userEvent.type(incrementAmount, '4');
```

File: babel.config.js
Comment (nickserv) [inline babel.config.js]: We should be able to simplify this by just using CRA's Babel config directly. It would also make our build tooling more consistent with how CRA would be used after generating the template.

```suggestion
  presets: ['react-app'],
```

**Review by nickserv — CHANGES_REQUESTED** [review]
Thanks, I had something similar in progress but this seems a bit more mature. 😄 

It would also be nice to update `App.test.js` so it's using `screen` for consistency with the new tests.

File: template/src/features/counter/counterSlice.test.js
Comment (aldreth) [reply template/src/features/counter/counterSlice.test.js:44]: @msutkowski I've changed these tests slightly & added comments to show the purpose. What do you think now?

File: babel.config.js
Comment (aldreth) [reply babel.config.js]: Good point

File: template/src/features/counter/fakeApi.js
Comment (aldreth) [reply template/src/features/counter/fakeApi.js]: It seems odd to call an api and then ignore what it returns, even for the purposes of demonstration.

Comment (BenLorantfy) [reply template/src/features/counter/fakeApi.js]: Could make a call to a random number api and increase the counter by what it returns? e.g. https://www.random.org/clients/http/api/

Comment (nickserv) [reply template/src/features/counter/fakeApi.js]: To be clear I'm suggesting actually using some fake data, not just making a request to make it async and ignoring the response.

Comment (aldreth) [reply template/src/features/counter/fakeApi.js]: Done - using https://www.random.org/clients/http/api/ @nickmccurdy @BenLorantfy

**Review by nickserv — APPROVED** [review]
Thanks, this is looking great!

File: template/src/features/counter/counterSlice.test.js
Comment (msutkowski) [reply template/src/features/counter/counterSlice.test.js:44]: Looks good to me, thanks!

**Comment by nickserv** [comment]
Is this ready to merge? Seems like it to me, unless anyone would like to leave a new review.

**Comment by aldreth** [comment]
Once it is merged I can put together a typescript version too.

**Pull request closed without merging**
